p2p: refactor AlreadyHave(), CInv::type, INV/TX processing

jonatack opened the pull request:
Building on #19590 and the recent `wtxid` and `GenTxid` changes, this is a refactoring and cleanup PR to simplify and improve some of the net processing code.

Some of the diffs are best reviewed with `-w` to ignore spacing.

Co-authored by John Newbery.

jonatack committed:
p2p: add CInv block message helper methods

jnewbery committed:
[net processing] Split AlreadyHave() into separate block and tx functions

jnewbery committed:
[net processing] Remove mempool argument from AlreadyHaveBlock()

jnewbery committed:
[net processing] Change AlreadyHaveBlock() to take block_hash argument

jnewbery committed:
[net processing] Change AlreadyHaveTx() to take a GenTxid

jonatack commented:
Rebased on merged #19590.

jonatack commented on src/protocol.h:
```diff
@@ -431,12 +431,20 @@ class CInv
+    bool IsMsgFilteredWitnessBlk() const { return type == MSG_FILTERED_WITNESS_BLOCK; }
 
     // Combined-message helper methods
-    bool IsGenTxMsg()     const { return type == MSG_TX || type == MSG_WTX || type == MSG_WITNESS_TX; }
```
@vasild per our offline convo, the helpers are now reverted to clang formatting ;)

fjahr reviewed (commented):
Code review ACK 82c92976792f9c9c11b38cb0092f18728c97e64e

jonatack commented on src/protocol.h:
```diff
@@ -431,12 +431,20 @@ class CInv
 
     // Combined-message helper methods
-    bool IsGenTxMsg()     const { return type == MSG_TX || type == MSG_WTX || type == MSG_WITNESS_TX; }
+    bool IsGenTxMsg() const { return type == MSG_TX || type == MSG_WTX || type == MSG_WITNESS_TX; }
```
Thanks Fabian. I mulled on doing this, but I think I'd rather keep the method definitions independent.

jnewbery commented:
I think a better approach than the changes within `AlreadyHave()` is to split that function up into `AlreadyHaveTx()` and `AlreadyHaveBlock()` since those are two totally separate concepts. @jonatack would you be interested in taking the commits from here: https://github.com/jnewbery/bitcoin/commits/2020-07-split-already-have? (see also https://github.com/bitcoin/bitcoin/pull/19569#discussion_r462887483)

jnewbery commented on src/protocol.h:
```diff
@@ -431,12 +431,20 @@ class CInv
-    bool IsGenTxMsg()     const { return type == MSG_TX || type == MSG_WTX || type == MSG_WITNESS_TX; }
+    bool IsGenTxMsg() const { return type == MSG_TX || type == MSG_WTX || type == MSG_WITNESS_TX; }
+    bool IsMsgBlkOrMsgWitnessBlk() const { return type == MSG_BLOCK || type == MSG_WITNESS_BLOCK; }
+    bool IsGenBlkMsg() const { return type == MSG_BLOCK || type == MSG_FILTERED_BLOCK || type == MSG_CMPCT_BLOCK || type == MSG_WITNESS_BLOCK; }
```
Should this contain `MSG_FILTERED_WITNESS_BLOCK`?

jonatack replied:
Well-spotted; added it in the rebase for now. The one place this helper is currently used (https://github.com/bitcoin/bitcoin/pull/19610/files#diff-eff7adeaec73a769788bb78858815c91L1763) did not include `MSG_FILTERED_WITNESS_BLOCK`, so we'd need to be sure including it doesn't change behavior (or was an oversight).

jnewbery replied:
Oh, it looks like `MSG_FILTERED_WITNESS_BLOCK` isn't actually used and is 'reserved for future use': https://github.com/bitcoin/bips/blob/master/bip-0144.mediawiki#relay. I recommend removing it from here, removing `IsMsgFilteredWitnessBlk()` (since it's not used anywhere) and adding a comment to `MSG_FILTERED_WITNESS_BLOCK` saying it's unused (I think we could even comment out the value since it's unused in the code).

jnewbery reviewed (commented):
Concept ACK. Needs rebase.

jonatack committed:
p2p: use CInv block message helpers in net_processing.cpp

jonatack committed:
p2p: remove nFetchFlags from NetMsgType TX and INV processing

The nFetchFlags code can be removed here because GetFetchFlags() can only add
the MSG_WITNESS_FLAG, which is added to the CInv::type field. That CInv is only
passed to AlreadyHave() or ToGenTxid(), and neither of those functions do
anything different depending on whether the CInv type is MSG_TX or
MSG_WITNESS_TX.

Co-authored by: John Newbery <john@johnnewbery.com>

jonatack committed:
test: use CInv::MSG_WITNESS_TX flag in p2p_segwit

jonatack committed:
p2p: make gtxid(.hash) and fAlreadyHave localvars const

jonatack commented:
Rebased with four commits pulled in from @jnewbery's https://github.com/bitcoin/bitcoin/compare/master...jnewbery:2020-07-split-already-have branch as well as the changes in #19611.

laanwj commented:
> I think a better approach than the changes within AlreadyHave() is to split that function up into AlreadyHaveTx() and AlreadyHaveBlock() since those are two totally separate concepts.

Great idea. The block and transaction paths have drifted apart so much that it no longer makes sense to have a common function there.

laanwj commented on src/protocol.h:
```diff
@@ -400,9 +400,11 @@ enum GetDataMsg : uint32_t {
 /** inv message data */
 class CInv
 {
+private:
```
Personally I don't really know if it makes sense for "plain-data" network message types like CInv to have private fields.

jonatack replied:
Oh ok, in that case we may want to make CInv a struct to avoid confusion. As a class, it looked like encapsulation was desirable, and only one place in the code was still accessing `CInv::type` from the outside so it was easy to make it private.

jonatack replied:
Er, renaming `CInv` to `Inv` to make it a struct would be too much noise. Done, dropped the last commit "p2p: refactor CInv::type from public to private class member".

yufi1213 commented:
****

vasild commented on src/protocol.h:
```diff
@@ -412,14 +412,22 @@ class CInv
-    bool IsMsgWtx()       const { return type == MSG_WTX; }
+    bool IsUndefined() const { return type == UNDEFINED; }
+    bool IsMsgTx() const { return type == MSG_TX; }
+    bool IsMsgBlk() const { return type == MSG_BLOCK; }
```
[continuing conversation from [another PR](https://github.com/bitcoin/bitcoin/pull/19611#discussion_r463552637)]

nit: `s/Blk/Block/`

```sh
$ git grep -i blk src |wc -l
     108
$ git grep -i block src |wc -l
   11014
```

jonatack replied:
I prefer to pass on this for these reasons:
(a) to maintain a consistent naming style in CInv (`tx`, `blk`)
(b) `block` would make the methods names longer
(c) `blk` is an established abbreviation: in the codebase, block data files called `blk*.dat`, etc.
(d) the shorter `blk` seems more consistent with the abbreviated naming style in use in recent PRs like #19569

jnewbery commented on src/protocol.h:
```diff
@@ -412,14 +412,22 @@ class CInv
 
     // Combined-message helper methods
-    bool IsGenTxMsg()     const { return type == MSG_TX || type == MSG_WTX || type == MSG_WITNESS_TX; }
+    bool IsGenTxMsg() const { return type == MSG_TX || type == MSG_WTX || type == MSG_WITNESS_TX; }
```
Did you consider making these switch statements without defaults, so that the compiler warns if a new type is added to the enum but not here?

jonatack replied:
> Did you consider making these switch statements without defaults, so that the compiler warns if a new type is added to the enum but not here?

That works but requires turning `enum GetDataMsg` into an `enum class` or `enum struct`, which is a fairly noisy change. I'll propose and see what you think.

jonatack replied:
> Did you consider making these switch statements without defaults, so that the compiler warns if a new type is added to the enum but not here?

I think that requires turning `enum GetDataMsg` into an `enum class` or `enum struct`, which is a good but fairly noisy change. I might propose it and see what reviewers think.

jnewbery reviewed (commented):
One inline comment. A couple more suggestions:

- squash the _p2p: remove unused nFetchFlags from NetMsgType::INV processing_ and _p2p: remove unused nFetchFlags from NetMsgType::TX processing_ commits
- remove the _p2p: change CInv::type from int to uint32_t_ commit. I don't think it's useful here.

vasild commented on src/net_processing.cpp:
```diff
@@ -1771,7 +1761,7 @@ void static ProcessGetData(CNode& pfrom, const CChainParams& chainparams, CConnm
     if (it != pfrom.vRecvGetData.end() && !pfrom.fPauseSend) {
         const CInv &inv = *it++;
-        if (inv.type == MSG_BLOCK || inv.type == MSG_FILTERED_BLOCK || inv.type == MSG_CMPCT_BLOCK || inv.type == MSG_WITNESS_BLOCK) {
+        if (inv.IsGenBlkMsg()) {
```
Could this result in unintended behavioral change?

Previously if `inv.type == MSG_FILTERED_WITNESS_BLOCK` we would not execute the code inside this `if` block. With this patch we would execute it.

Related to https://github.com/bitcoin/bitcoin/pull/19610#discussion_r465175043

jonatack replied:
That message type is unused (see https://github.com/bitcoin/bitcoin/pull/19610#discussion_r468510147) and has since been commented out by #19721 (merged).

jonatack replied:
Removing it in the rebase.

vasild reviewed (commented):
In commits `p2p: remove unused nFetchFlags from NetMsgType::(TX|INV) processing` - why would it be unused? It looks used to me. Maybe I am missing something obscure? If that is the case, then maybe the commit messages warrant some more elaborate explanations?

The change
```diff
-            } else {
+            } else if (inv.IsGenTxMsg()) {
```
does not belong to commit `[net processing] Remove mempool argument from AlreadyHaveBlock()`

jnewbery commented:
> In commits p2p: remove unused nFetchFlags from NetMsgType::(TX|INV) processing - why would it be unused? It looks used to me. Maybe I am missing something obscure? If that is the case, then maybe the commit messages warrant some more elaborate explanations?

Because `GetFetchFlags()` can only add the `MSG_WITNESS_FLAG`, which is added to the `CInv.type` field. That `CInv` only passed to `AlreadyHave()` or `ToGenTxid()` and neither of those functions do anything different depending on whether the `CInv` type is `MSG_TX` MSG_WITNESS_TX`. I agree that the git commit log could be expanded to explain this.

> The change [...] does not belong to commit [net processing] Remove mempool argument from AlreadyHaveBlock()

It's an important change. See https://github.com/bitcoin/bitcoin/pull/19569#discussion_r462887483 for context. Perhaps it should be split into a separate commit or have a commit log explaining why it's needed.

jnewbery commented:
Needs rebase now that #19721 is merged.

jnewbery commented:
Still awaiting rebase

jonatack commented:
- rebased and updated
- dropped the *change CInv::type from int to uint32_t commit*
- squashed the 2 commits *remove unused nFetchFlags from NetMsgType::INV processing* and *remove unused nFetchFlags from NetMsgType::TX processing* and added explanatory notes to the commit message
- split the IsGenMsgTx() conditional check to a separate commit, *ensure inv is GenMsgTx before ToGenTxid in inv processing*, and added explanatory notes in the commit message

jnewbery committed:
p2p: ensure inv is GenMsgTx before ToGenTxid in inv processing

and otherwise log that an unknown INV type was received.

In INV processing, when handling transaction type inv messages,
ToGenTxid() expects that we constructed the CInv ourselves or
that we verified that it is for a transaction type CInv.

Therefore, change this `else` branch into an `else if (inv.GenMsgTx())`
to make this safer and log any INVs that fall through.

jnewbery reviewed (commented):
utACK 2a2ee28a2ac2d2cc11ad8a44b036bb4218c13d9a

One suggestion for improving logging inline.

jnewbery commented:
utACK 25176e62d0e60a3bca92aa758f935ee782531f75

jonatack commented:
Rebased.

`git range-diff 7f609f6 25176e6 ac0617d`

jnewbery commented:
utACK ac0617d3718ac3e330a70b30ef1edd8a46faaf9a

Verified the range-diff locally

vasild commented on src/protocol.h:
```diff
@@ -418,12 +418,24 @@ class CInv
     // Single-message helper methods
-    bool IsMsgTx()        const { return type == MSG_TX; }
-    bool IsMsgWtx()       const { return type == MSG_WTX; }
+    bool IsUndefined() const { return type == UNDEFINED; }
```
nit: the methods `IsUndefined()` and `IsMsgWitnessTx()` are unused.

jonatack replied:
Good point; probably worth removing them for now.

jonatack replied:
Done

vasild reviewed (approved):
ACK ac0617d37

Looks good, compiled locally and ran tests.

Crypt-iQ commented:
I think the `CInv::type` change from `int` to `uint32_t` change should be added back (not necessarily the `private` qualifier). While fuzzing it brings up the following complaint https://github.com/bitcoin/bitcoin/issues/19678

jonatack commented:
> I think the `CInv::type` change from `int` to `uint32_t` change should be added back (not necessarily the `private` qualifier). While fuzzing it brings up the following complaint #19678

Thank you, @Crypt-iQ, interesting. The type suggestion was originally by @vasild in https://github.com/bitcoin/bitcoin/pull/19590#pullrequestreview-455788826; will try fuzzing with `uint32_t`.

jonatack commented:
Thank you @Crypt-iQ, if you have the chance, would you please let me know if 533a9a0d2c307330497e61d5af4229c924db8372 resolves the issue on your side?

Crypt-iQ commented:
@jonatack That UBSan complaint goes away with this patch 👍 
Tested on macOS v10.15.4, brew-installed clang.

vasild reviewed (approved):
ACK 533a9a0d2

I can't make sense of the CI failure (and tell whether it is related to this PR).

jonatack commented:
Thanks @Crypt-iQ and @vasild.

The CI failed on the wallet_import_rescan functional test and should be unrelated--restarted it to see.

jonatack commented:
Thanks @Crypt-iQ and @vasild.

The CI failed on the wallet_import_rescan functional test and should be unrelated. Restarted; it's now green.

jonatack commented:
Removed the unused helper methods from the first commit.

vasild reviewed (approved):
ACK 125468af9

jnewbery commented:
> I think the CInv::type change from int to uint32_t change should be added back (not necessarily the private qualifier). While fuzzing it brings up the following complaint #19678

Why? This seems unrelated to this PR (the UBSan warning happens before this PR). Changing the type of `CInv.type` (which is serialized/deserialized for network messages) is a very different change from the internal refactors in this PR. Can you save it for a different PR?

Crypt-iQ commented:
@jnewbery I thought since it was already added previously, it would fit into the scope of the PR. No problem to be added into another PR, either I or somebody else could do that.

jonatack commented:
Dropped the `CInv::type` type change commit in favor of #19818. Thanks everyone.

laanwj commented:
Code review ACK fb56d37612dea6666e7da73d671311a697570dae

jnewbery commented:
utACK fb56d37612dea6666e7da73d671311a697570dae

vasild reviewed (approved):
ACK fb56d3761
